### DISCUSSION TITLE
MDX blog just displays markdown content instead of rendering it while using `MDXProvider` from `@mdx-js/react` in Next JS

### DISCUSSION THREAD

**Discussion by deadcoder0904** [vercel/next.js#17283] [Help] [answered]
I am making a blog with [MDX](https://mdxjs.com) & [Next.js](https://nextjs.org) but I am unable to render Markdown content. The blog post just shows markdown content as `string`.

Here's my complete source code → https://github.com/deadcoder0904/blog-mdx-next/

I have the following folder structure:

```bash
.
|-- README.md
|-- components
|   `-- Image.js
|-- next.config.js
|-- package-lock.json
|-- package.json
|-- pages
|   |-- _app.js
|   |-- blog
|   |   `-- [slug].js
|   |-- dark.css
|   |-- index.js
|   `-- new.css
|-- posts
|   |-- blog
|   |   |-- hello-world
|   |   |   |-- Rustin_Cohle.jpg
|   |   |   `-- index.mdx
|   |   `-- shit-world
|   |       `-- index.mdx
|   `-- tutorials
|       `-- console-log-in-javascript
|           `-- index.mdx
|-- prettier.config.js
`-- utils
    `-- mdxUtils.js
```

I have all my content in `posts/` folder.

I have 2 folders in it: `blog/` & `tutorials/`

Each post is in their own folder inside of `blog/` or `tutorials/`. It looks like:

[![blog archives][1]][1]

I want to render my `posts/blog/hello-world/index.mdx` post at `blog/hello-world` location so I made `blog/[slug].js` file.

The contents of it are as follows:

# pages/blog/[slug].js

```js
import fs from 'fs'
import path from 'path'
import matter from 'gray-matter'
import { MDXProvider } from '@mdx-js/react'

import { BLOG_PATH, blogFilePaths } from '../../utils/mdxUtils'
import { Image } from '../../components/Image'

const MDXComponents = { Image }

const Blog = ({ source, frontMatter }) => {
	return (
		<div>
			<h1>{frontMatter.title}</h1>
			<MDXProvider components={MDXComponents}>{source}</MDXProvider>
		</div>
	)
}

export async function getStaticPaths() {
	const paths = blogFilePaths.map((path) => {
		const split = path.split('/')
		const slug = split[split.length - 2]
		return {
			params: {
				slug,
			},
		}
	})

	return {
		paths,
		fallback: false,
	}
}

export const getStaticProps = async ({ params }) => {
	const { slug } = params
	const blogFilePath = path.join(BLOG_PATH, `/blog/${slug}/index.mdx`)

	const source = fs.readFileSync(blogFilePath)
	const { content: mdx, data } = matter(source)

	if (!blogFilePath) {
		console.warn('No MDX file found for slug')
	}

	return {
		props: {
			source: mdx,
			frontMatter: data,
		},
	}
}

export default Blog
```

The important part is:

```js
<MDXProvider components={MDXComponents}>{source}</MDXProvider>
```

I thought this would render markdown content but it only displays markdown content as a `string`.

You can check the output at https://codesandbox.io/s/github/deadcoder0904/blog-mdx-next?file=/pages/blog/%5Bslug%5D.js by clicking any blog post.

It displays the following when I click on `Hello World` post:

[![hello world][2]][2]

How do I actually render the content?

I looked at other Github repos like [Tailwind CSS Blog](https://github.com/tailwindlabs/blog.tailwindcss.com) & it works [fine for them](https://blog.tailwindcss.com/) but I am not sure how it works sadly :(

I do know that I have to convert the `source` prop in `Blog` or `mdx` in `getStaticProps` but I [don't see](https://github.com/tailwindlabs/blog.tailwindcss.com/blob/master/src/components/Post.js#L88) Tailwind doing it either.

  [1]: https://i.stack.imgur.com/2X5bE.png
  [2]: https://i.stack.imgur.com/E2BJc.png

**Comment by lfades** [comment, marked as answer]
@deadcoder0904 This is not a bug but expected behavior instead.

For `MDXProvider` to do something you need to use React elements as children. Right now the children is a string, and it will be treated as so.

Take a look at [`next-mdx-remote`](https://github.com/hashicorp/next-mdx-remote) instead, it will allow you to use your custom components 👍

**Reply by deadcoder0904** [reply]
I already used `next-mdx-remote` but switched from it because I cannot use images from the same folder as the blog post.

More info about it → https://stackoverflow.com/questions/63957018/how-to-use-images-in-a-mdx-file-outside-of-public-folder-while-using-next

> For MDXProvider to do something you need to use React elements as children. Right now the children is a string, and it will be treated as so.

What do you mean by the children is a string?

I tried doing:

```js
<MDXProvider components={MDXComponents}>
	<source />
</MDXProvider>
```

But it didn't work.

I also tried looking at https://github.com/hashicorp/next-mdx-remote/blob/master/render-to-string.js so as to use it in my app but it didn't work either.

Can you give an example since I've been stuck with this for a long time?

**Reply by lfades** [reply]
You need to render a React Element, not a string. There's a big difference between this:

```
# Hello World
```

And this:

```jsx
// jsx
<h1>Hello World</h1>

// Which transforms into
React.createElement('h1', null, 'Hello World')
```

**Reply by lfades** [reply]
@deadcoder0904 And about that error you had with next-mdx-remote. I'll be doing a PR to that package soon for some improvements

**Reply by deadcoder0904** [reply]
> You need to render a React Element, not a string.

Yes, I got that. But I'm unable to solve it as Idk what I need to do. There's a similar project that I found through searching on Github → https://github.com/uu64/nextjs-mdx-blog-sample but it uses glob array (babel plugin) to render `.mdx` file. If you know a concrete answer would definitely love some help as I've exhausted all my resources :)

> I'll be doing a PR to that package soon for some improvements

Would love to see that. I'll be keeping an eye on it as it made my code a heck of a lot simpler but do ping me if you remember. Although, the author said that it can't be done technically as the package escapes `webpack` & treats markdown content as data. But would love it as it's such a basic use-case to build a blog with images in their own folder.

**Comment by tconroy** [comment]
Hey @deadcoder0904 - I'm trying to do the same as you with a very similar setup -- seeing my markdown rendered as a string instead of executing as HTML / react component(s). Did you ever figure out a solution to this? Any help much appreciated!

**Reply by deadcoder0904** [reply]
@tconroy I answered you on [StackOverflow](https://stackoverflow.com/questions/64007838/mdx-blog-just-displays-markdown-content-instead-of-rendering-it-while-using-mdx/64055028?noredirect=1#comment114967013_64055028) :)
